Feature/high resolution timestamps

reftel opened the pull request:
Not sure "Timestamp" is the best choice of name, as it's kind of overloaded, but I didn't come up with anything better.

reftel committed:
Test parsing and formatting of timestamps

reftel committed:
Allow for higher-precision timestamps

Resolves issue #365

grandinj commented:
ok, that's an interesting way to solve the problem :-)

There will probably be some time drift, and it won't track system clock changes. Not sure how much of a problem that is likely to be in practice.

I don't think we need another Timestamp class, we can just pass around long values containing microsecond level accuracy UTC time. 64-bits is enough for that.

reftel commented:
Right, this won't get any NTP corrections. That's not good. I'll see if I can come up with something better.

svladykin commented:
Commented at the respective issue #365

reftel committed:
test: Close connection after TestFunctions.testOraHash

reftel committed:
test: Close connections in timestamp tests

reftel commented:
Updated to add a `TIME_SOURCE` option that lets uses choose between the traditional behaviour (which is the default), start time plus nanoseconds since start, or some user-defined behaviour.
Updated the documentation to reflect that the argument to `CURRENT_TIMESTAMP` is for resolution and not precision.

svladykin commented:
@reftel I don't really like your naming: `AccurateButImprecise` and like this. It looks weird and tells your opinion about the implementation, but does not objectively describe it. For me it may be precise enough. So if your implementation uses `System.currentTimeMillis()`, then you can name it like `SystemMillisecondTimeSource`. All the side effects you can describe in javadoc.

svladykin commented:
@reftel Also please make sure that you code conforms to checkstyle. I saw some weirdly formatted javadocs in new classes, may be there are some other code style problems.

reftel commented:
I ran checkstyle on it, and found nothing in my changes. I'll change the Javadoc formatting to have the description paragraph on a new line, though.
The naming reflects the tradeoffs and not the implementation, but OK. `MillisecondTimeSource` for the `System.currentTimeMills` one, but what about the other? Just `NanosecondTimeSource` would make it sound like it is just better, without warning the user of the side effects. `ElapsedTimeSource`?

reftel commented:
Javadoc reformatted. Naming not changed yet.

svladykin commented:
I guess something like `ElapsedNanoTimeSource` must be ok.

reftel commented:
Renamed to `org.h2.time.MillisecondTimeSource` and `org.h2.time.ElapsedNanoTimeSource`

grandinj commented:
Really nice work, but I still don't think we need a new Timestamp class. Just pass around long values containing microsecond level accuracy values.
When we have terahertz CPUs with nanosecond level accuracy atomic-clocks we can revisit that decision.

reftel commented:
@svladykin Not sure I really like the new naming after all. How about org.h2.time.Accurate and org.h2.time.Precise?

@grandinj I think it's confusing enough that longs sometimes contain seconds and sometimes milliseconds. Adding another weakly typed measure of time just adds to the confusion.

grandinj commented:
@reftel I'm not keen on adding yet another Timestamp type to our existing rather confusing menagerie 

TBH, I'm not sure this feature is worth it's weight anymore. It adds quite a lot of extra stuff for something that really does not seem that useful (to me)

Sorry, but I think I'll leave this one up to @svladykin to decide or if @thomasmueller  shows up, he can make the call.

svladykin commented:
@reftel Sorry for my poor English but I can't even distinguish between them, even Google Translate translates `Accurate` and `Precise` to Russian as the same word :)

It is not the only issue with naming like this. Suppose we decided to drop one of them, does the other name will make sense then? I'm not sure. Or suppose we want to add one more implementation, how would we call it? `VeryVeryPrecise`? `NotThatMuchAccurate`? Looks bad naming convention to me.

This is why I insist on having implementations of an interface named depending on what they are actually doing, but not on presence or absence of other implementations or your opinions about their accuracy or precision. I noticed your earlier argument about tradeoffs, but it is wrong here since you do not know the tradeoffs for all the future changes. Makes sense?

@grandinj Ok, I'll try to figure out what we can do here to make it cleaner.

reftel commented:
@svladykin I see your point. If we add more standard time sources, when naming them could be a problem (dropping one would be a breaking change, so we could probably rename the remaining ones if we do that). So let's be specific in what source for time the `TimeSource`s use. How about `SystemCurrentTimeMillis` and `SystemNanoTime`?

I agree that the difference between accurate and precise can be subtle (and apparently not just for us non-native speakers). There's some discussion at https://en.wikipedia.org/wiki/Accuracy_and_precision , and Java seems to use the terms similarly. For instance, the Javadoc for `System.nanoTime` says "This method provides nanosecond precision, but not necessarily nanosecond accuracy".

@grandinj Would it help if the class was called something other than `Timestamp`?

W.r.t. the usefulness, it varies a lot with the application. For most applications, millisecond precision is more than enough. For the application feature that motivated this change, only having millisecond precision is a deal breaker.

svladykin commented:
`SystemCurrentTimeMillis` and `SystemNanoTime` look good to me. 

As for `Timestamp` I don't have a clear opinion right now, let me take a closer look at your PR on the weekend.

svladykin commented:
@reftel Sorry for a long delay, was a bit busy.

I looked at the code. I tend to agree with Noel, that adding new type `Timestamp` does not make much sense if we can put all the meaningful precision to a primitive `long`.

But to avoid any confusions about time units we should do the following:  
rename fields `sessionStart` and `transactionStart` to `sessionStartMicros` and `transactionStartMicros` respectively and add explicit `TimeUnit` parameter to the accessor methods.

What do you think?
